cmd/cloud-api-adaptor/cloudmgr/manager.go

```diff
@@ -0,0 +1,38 @@
+// (C) Copyright Confidential Containers Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package cloudmgr
+
+import (
+	"flag"
+	"os"
+
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/cloud"
+)
+
+type Cloud struct {
+	ParseCmd    func(flags *flag.FlagSet)
+	LoadEnv     func()
+	NewProvider func() (cloud.Provider, error)
+}
+
+var cloudTable map[string]*Cloud = make(map[string]*Cloud)
```

Comment:
As only one provider will be available per binary. Creating a one entry map seems redundant, I'd suggest the following approch instead. Code shorten (let me know if you would like more details)

manager.go
```go
package cloudmgr

type Cloud interface {
	ParseCmd(flags *flag.FlagSet)
	LoadEnv()
	NewProvider() (cloud.Provider, error)
}
```
ibmcloud.go
```go
//go:build ibmcloud
package cloudmgr
type IBMCloudMgr struct{}

func Get() Cloud {
	return IBMCloudMgr{}
}
func (IBMCloudMgr) ParseCmd(flags *flag.FlagSet) { ... }
func (IBMCloudMgr) LoadEnv() { ... }
func (IBMCloudMgr) NewProvider() (cloud.Provider, error) { ... }
```

Likewise following providers would follow pattern

aws.go
```go
//go:build aws
package cloudmgr
type AWSCloudMgr struct{}
func Get() Cloud {
	return AWSCloudMgr{}
}
etc...
```

Reply:
Thank you for your comment. Let me explain my intention here.

My intention is to add ability to support multiple providers in a single binary. We can still switch cloud provider by specifying a cloud provider name in a first command line argument, as we do in the current implementation.

When we build a binary with multiple build tags like `go build -tags ibmcloud,aws ...`, the output binary will support both ibmcloud and aws.  When we build a binary with a single build tag like `go build -tags ibmcloud ...`, the output binary will support ibmcloud only.

Reply:
I didn't think we'd support running multiple providers from the same binary? I'm not sure I can see a scenario where that would be used.

Reply:
How about discussing the pros/cons of supporting multiple providers in a single binary via a separate github issue and if required in the community interlock ?

Reply:
OK. I'll raise a new issue for discussing this point.

Reply:
I have changed the struct to interface, but I couldn't remove the `cloudTable` map. I am currently refactoring the ibmcloud provider code only, and not touching the other providers. 

If I define `Get()` with the `ibmcloud` build tag, CI tests for the other fail for now, since I don't prepare `Get()` methods in the other providers.

I'll change this to the proposed code, when all of the cloud providers are refactored.

Reply:
I tried to remove the `cloudTable` map, and suppress golint errors by editing `.golangci.yaml`, but I can suppress the following errors.

```
golangci-lint (1.18): cmd/cloud-api-adaptor/cloudmgr/azure.go#L8
`Get` redeclared in this block (typecheck)
golangci-lint (1.18): cmd/cloud-api-adaptor/cloudmgr/aws.go#L8
other declaration of Get (typecheck)
golangci-lint (1.18): cmd/cloud-api-adaptor/cloudmgr/ibmcloud.go#L15
`Get` redeclared in this block (typecheck)
golangci-lint (1.18): cmd/cloud-api-adaptor/cloudmgr/libvirt.go#L8
`Get` redeclared in this block (typecheck)
```

For now, I continue to use  the `cloudTable` map, and I will come back here when the other cloud provider refactoring is done.

Reply:
I raise an issue for this topic.
 #573